File: Prism/Translation/LanguageCode.swift
import Foundation

struct LanguageCode: Identifiable, Hashable {
    let id: String
    let displayName: String

    static let commonTargets: [LanguageCode] = [
        .init(id: "en", displayName: "English"),
        .init(id: "pt", displayName: "Português"),
        .init(id: "es", displayName: "Español"),
        .init(id: "fr", displayName: "Français"),
        .init(id: "de", displayName: "Deutsch"),
        .init(id: "it", displayName: "Italiano"),
        .init(id: "ja", displayName: "日本語"),
        .init(id: "ko", displayName: "한국어"),
        .init(id: "zh-Hans", displayName: "中文 (简体)"),
        .init(id: "zh-Hant", displayName: "中文 (繁體)"),
        .init(id: "ru", displayName: "Русский"),
        .init(id: "ar", displayName: "العربية"),
        .init(id: "nl", displayName: "Nederlands"),
        .init(id: "pl", displayName: "Polski"),
        .init(id: "tr", displayName: "Türkçe"),
        .init(id: "hi", displayName: "हिन्दी")
    ]

    static func displayName(for id: String) -> String {
        let normalized = normalize(id)
        return commonTargets.first(where: { $0.id == normalized })?.displayName
            ?? Locale.current.localizedString(forLanguageCode: normalized)
            ?? id
    }

    /// Ex.: "detecção automática → English" ou "Português → English".
    static func pairLabel(from: String?, to: String) -> String {
        let source = from.map { displayName(for: $0) } ?? "detecção automática"
        return "\(source) → \(displayName(for: to))"
    }

    /// Maps provider codes (EN, PT-BR, zh-CN, …) onto app language ids when possible.
    static func normalize(_ code: String) -> String {
        let lower = code.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
            .replacingOccurrences(of: "_", with: "-")
        switch lower {
        case "en", "en-us", "en-gb": return "en"
        case "pt", "pt-br", "pt-pt": return "pt"
        case "zh", "zh-cn", "zh-hans", "zh-hans-cn": return "zh-Hans"
        case "zh-tw", "zh-hk", "zh-hant", "zh-hant-tw": return "zh-Hant"
        default:
            if commonTargets.contains(where: { $0.id == lower }) { return lower }
            let base = Locale(identifier: lower).language.languageCode?.identifier ?? lower
            if commonTargets.contains(where: { $0.id == base }) { return base }
            return lower
        }
    }

    /// Idioma preferido do sistema, mapeado para um código suportado pelo app.
    static var systemLanguage: String {
        let preferred = Locale.preferredLanguages.first ?? "en"
        let code = Locale(identifier: preferred).language.languageCode?.identifier ?? "en"
        if commonTargets.contains(where: { $0.id == code }) {
            return code
        }
        // Map regional variants
        if code == "zh" {
            return preferred.lowercased().contains("hant") || preferred.lowercased().contains("tw")
                ? "zh-Hant" : "zh-Hans"
        }
        return "en"
    }

    /// Destino padrão: inglês (ou português se o sistema já for inglês), para não coincidir com a origem.
    static var systemDefaultTarget: String {
        systemLanguage == "en" ? "pt" : "en"
    }
}

File: Prism.Windows/src/Prism.Core/Translation/LanguageCode.cs
namespace Prism.Core.Translation;

public sealed record LanguageCode(string Id, string DisplayName)
{
    public string id => Id;

    public static IReadOnlyList<LanguageCode> CommonTargets { get; } =
    [
        new("en", "English"),
        new("pt", "Português"),
        new("es", "Español"),
        new("fr", "Français"),
        new("de", "Deutsch"),
        new("it", "Italiano"),
        new("ja", "日本語"),
        new("ko", "한국어"),
        new("zh-Hans", "中文 (简体)"),
        new("zh-Hant", "中文 (繁體)"),
        new("ru", "Русский"),
        new("ar", "العربية"),
        new("nl", "Nederlands"),
        new("pl", "Polski"),
        new("tr", "Türkçe"),
        new("hi", "हिन्दी"),
    ];

    public static string DisplayNameFor(string id)
    {
        var normalized = Normalize(id);
        var known = CommonTargets.FirstOrDefault(c => c.Id == normalized);
        if (known is not null) return known.DisplayName;

        try
        {
            var culture = new System.Globalization.CultureInfo(normalized);
            return culture.DisplayName;
        }
        catch (System.Globalization.CultureNotFoundException)
        {
            return id;
        }
    }

    public static string PairLabel(string? from, string to)
    {
        var source = from is null ? "automatic detection" : DisplayNameFor(from);
        return $"{source} → {DisplayNameFor(to)}";
    }

    public static string Normalize(string code)
    {
        var lower = code.Trim().ToLowerInvariant().Replace('_', '-');
        return lower switch
        {
            "en" or "en-us" or "en-gb" => "en",
            "pt" or "pt-br" or "pt-pt" => "pt",
            "zh" or "zh-cn" or "zh-hans" or "zh-hans-cn" => "zh-Hans",
            "zh-tw" or "zh-hk" or "zh-hant" or "zh-hant-tw" => "zh-Hant",
            _ when CommonTargets.Any(c => c.Id == lower) => lower,
            _ => TryBaseLanguage(lower),
        };
    }

    private static string TryBaseLanguage(string lower)
    {
        try
        {
            var baseCode = new System.Globalization.CultureInfo(lower).TwoLetterISOLanguageName;
            if (CommonTargets.Any(c => c.Id == baseCode))
                return baseCode;
        }
        catch (System.Globalization.CultureNotFoundException)
        {
            // fall through
        }

        var dash = lower.IndexOf('-');
        if (dash > 0)
        {
            var basePart = lower[..dash];
            if (CommonTargets.Any(c => c.Id == basePart))
                return basePart;
        }

        return lower;
    }

    public static string SystemLanguage
    {
        get
        {
            var preferred = System.Globalization.CultureInfo.CurrentUICulture;
            var code = preferred.TwoLetterISOLanguageName;
            if (CommonTargets.Any(c => c.Id == code))
                return code;

            if (code == "zh")
            {
                var name = preferred.Name.ToLowerInvariant();
                return name.Contains("hant") || name.Contains("tw") || name.Contains("hk")
                    ? "zh-Hant"
                    : "zh-Hans";
            }

            return "en";
        }
    }

    public static string SystemDefaultTarget => SystemLanguage == "en" ? "pt" : "en";
}
